SamFileHeader was not available causing a NPE

nh13 opened the pull request:
- [ ] I am running this on a test sample to verify it runs to completion.  I will update this PR once it has succeeded.

yfarjoun commented:
I had heard complaints that this code doesn't work. @takutosato here's the bug fix! The fact that this wasn't tested at all is troubling. Could you add a test that just runs the code against a small bam and verifies that it creates a metric?

takutosato commented:
@yfarjoun will do.

takutosato commented:
The tools still doesn't run; you get an empty header inside `getIntervalsToExamine()` (the line above super.doWork(). I have a fix though. PR coming in the next day or two.

nh13 commented:
@takutosato I have an incoming fix for that too, so I will append a few tests and fixes to this PR.

nh13 commented:
@takutosato and @yfarjoun I have fixed a number of issues and added some basic tests.  Again, I am going to run this on a large dataset to verify the results of these fixes.

yfarjoun commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -480,6 +487,7 @@ protected IntervalList getIntervalsToExamine() {
     }
 
+    /** This method should only be called after {@link this.getSamReader()} is called. */
     protected SAMFileHeader getSamFileHeader() {
```
add an exception if not?

yfarjoun commented on src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java:
```diff
@@ -0,0 +1,131 @@
+        for (final Histogram<Integer> histogram : output.getAllHistograms()) {
+            if (histogram.getValueLabel().equals("count_WHOLE_GENOME")) {
+                Assert.assertEquals(histogram.get(0).getValue(), 274d);
+            }
```
} else {

😄

yfarjoun commented on src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java:
```diff
@@ -0,0 +1,131 @@
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 0.107438);
+            }
```
here

yfarjoun commented on src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java:
```diff
@@ -0,0 +1,131 @@
+        for (final Histogram<Integer> histogram : output.getAllHistograms()) {
+            if (histogram.getValueLabel().equals("count_WHOLE_GENOME")) {
+                Assert.assertEquals(histogram.get(0).getValue(), 1080d);
+            }
```
here

yfarjoun commented on src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java:
```diff
@@ -0,0 +1,131 @@
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 0.321782);
+            }
```
here

yfarjoun commented:
still minor comments, otherwise good-to-go!

nh13 commented:
@yfarjoun I wanted to wait for #658 but on the other hand I'd like to get this in this week (before a release) due to the tool not being useful in its current state.  Given the volume of comments on #658 I'd like to fix this up and merge it.  Thoughts?

yfarjoun commented:
I think that #658 will still take a while to review and merge...so I think that you can go ahead with fixing this PR. I'd rather have the tool in working condition. The only overlap is the actual solution to the NPE..

takutosato commented:
Yea, @nh13 please consider my solution to the NPE in #658 (i.e. override `getCollector` in NonZeroCoverage) and choose whichever solution you like. Thanks for giving me a chance to merge first. I wasn't able to deliver but I appreciate it.

yfarjoun commented:
@nh13 I had made comments on this PR a while ago and it was merged with no reference to my comments... 😕

nh13 commented:
Minor comments fixed so then good to go. See your previous comment. I took that as fix it up then merge.

yfarjoun commented:
but you didn't fix the comments I raised! the `} else {` comments and also the request that you throw an exception if an assumption isn't met....

nh13 commented:
I see. I can make a quick PR this weekend. What was I thinking?

yfarjoun commented:
thanks. 👍
